src/base1.rs

```diff
@@ -0,0 +1,1947 @@
+                // we have removed.
+                //
+                // By unlinking nodes in batches we make sure that the final search doesn't
+                // unlink all nodes at once, which could keep the current thread pinned for a
+                // long time.
+                let mut entry = self.lower_bound(Bound::Unbounded, guard);
+
+                for _ in 0..BATCH_SIZE {
+                    // Stop if we have reached the end of the list.
+                    let e = match entry {
+                        None => return,
+                        Some(e) => e,
+                    };
+
+                    // Before removing the current entry, first obtain the following one.
+                    let next = e.next();
+
+                    // Try removing the current entry.
+                    if e.node.mark_tower() {
+                        // Success! Decrement `len`.
+                        self.hot_data.len.fetch_sub(1, Ordering::Relaxed);
+                    }
+
+                    entry = next;
+                }
+            }
+
+            // Repin the current thread because we don't want to keep it pinned in the same
+            // epoch for a too long time.
+            guard.repin();
+        }
+    }
+}
+
+impl<K, V> Drop for SkipList<K, V> {
+    fn drop(&mut self) {
+        unsafe {
+            let mut node = self.head[0]
+                .load(Ordering::Relaxed, epoch::unprotected())
+                .as_ref();
+
+            // Iterate through the whole skip list and destroy every node.
+            while let Some(n) = node {
+                // Unprotected loads are okay because this function is the only one currently using
+                // the skip list.
+                let next = n.tower[0]
+                    .load(Ordering::Relaxed, epoch::unprotected())
+                    .as_ref();
+
+                // Deallocate every node.
+                Node::finalize(n);
+
+                node = next;
+            }
+        }
+    }
+}
+
+impl<K, V> IntoIterator for SkipList<K, V> {
+    type Item = (K, V);
+    type IntoIter = IntoIter<K, V>;
+
+    fn into_iter(self) -> IntoIter<K, V> {
+        unsafe {
+            // Load the front node.
+            //
+            // Unprotected loads are okay because this function is the only one currently using
+            // the skip list.
+            let front = self.head[0]
+                .load(Ordering::Relaxed, epoch::unprotected())
+                .as_raw();
+
+            // Clear the skip list by setting all pointers in head to null.
+            for level in 0..MAX_HEIGHT {
+                self.head[level].store(Shared::null(), Ordering::Relaxed);
+            }
+
+            IntoIter {
+                node: front as *mut Node<K, V>,
+            }
+        }
+    }
+}
+
+/// An entry in a skip list, protected by a `Guard`.
+pub struct Entry<'a, 'g, K: 'a + 'g, V: 'a + 'g> {
```

Comment:
Is there any point in having two lifetimes? :) Why not define `front` and other methods like this:

```rust
fn front<'a>(&'a self, guard: &'a Guard) -> Option<Entry<'a, K, V>>
```

Reply:
The guard and self lifetimes needs to be separate for `pin` to work.